File: 1146-greatest-common-divisor-of-strings/1146-greatest-common-divisor-of-strings.cpp
class Solution {
public:
    string gcdOfStrings(string str1, string str2) {
        int n1=str1.length();
        int n2=str2.length();
        if((str1+str2) == (str2+str1)){
            int g= gcd(n1, n2);
            return str1.substr(0, g);
        } else{ return "";}
    }

    int gcd(int x, int y){
        if(x==0){ return y;}
        if(y==0){ return x;}
        if(x==y){return x;}
        if(x>y){
            return gcd(x-y, y);
        }
        return gcd(x, y-x);
    }
};

File: 1146-greatest-common-divisor-of-strings/1146-greatest-common-divisor-of-strings.java
class Solution {
    public String gcdOfStrings(String str1, String str2) {
        if((str1+str2).equals(str2+str1)){
            int gcd = gcd(str1.length(), str2.length());
            return str1.substring(0, gcd);
        }
        return "";
    }
    public int gcd(int a, int b){
        if(a==0) return b;
        if(b==0) return a;
        if(a>b){ return gcd(a-b, b);}
        return gcd(a, b-a);
    }
}